front_end/ndb/NdbMain.js

```diff
@@ -255,53 +273,72 @@ Ndb.NodeProcessManager = class extends Common.Object {
   }
 
   async _sendMesage(message) {
-    const service = await this._service();
-    return service.sendMessage(message);
-  }
-
-  _service() {
-    if (!this._servicePromise) {
-      async function service() {
-        const service = await Ndb.backend.createService('ndd_service.js');
-        await service.init(rpc.handle(this));
-        InspectorFrontendHost.sendMessageToBackend = this._sendMesage.bind(this);
-        return service;
-      }
-      this._servicePromise = service.call(this);
-    }
-    return this._servicePromise;
+    return this._service.sendMessage(message);
   }
 
   async _onExecutionContextDestroyed(event) {
     const executionContext = event.data;
-    const mainContextId = 1;
-    if (executionContext.id !== mainContextId)
+    if (!executionContext.isDefault)
       return;
     const target = executionContext.target();
-    if (target.suspended()) {
-      const debuggerModel = target.model(SDK.DebuggerModel);
-      await new Promise(resolve => debuggerModel.addEventListener(
-          SDK.DebuggerModel.Events.DebuggerWasEnabled, resolve));
+    if (this._isProfiling) {
+      this._cpuProfiles.push({
+        profile: await target.model(SDK.CPUProfilerModel).stopRecording(),
+        name: target.name(),
+        id: target.id()
+      });
     }
-    const service = await this._service();
-    service.disconnect(target.id());
+    await this._service.disconnect(target.id());
   }
 
   async debug(execPath, args, options) {
     options = options || {};
-    const service = await this._service();
     const debugId = options.data || String(++this._lastDebugId);
-    this._lastStarted = {execPath, args, debugId};
+    if (!options.data)
+      this._lastStarted = {execPath, args, debugId, isProfiling: this._isProfiling};
 
     const {cwd} = await Ndb.processInfo();
-    return service.debug(
+    return this._service.debug(
         execPath, args, {
           ...options,
           data: debugId,
           cwd: cwd,
         });
   }
 
+  async profile(execPath, args, options) {
+    await UI.viewManager.showView('timeline');
+    const action = UI.actionRegistry.action('timeline.toggle-recording');
+    await action.execute();
+    this._isProfiling = true;
+    await this.debug(execPath, args);
+    this._isProfiling = false;
+    this._cpuProfiles.push(...await Promise.all(SDK.targetManager.models(SDK.CPUProfilerModel).map(async profiler => ({
+      profile: await profiler.stopRecording(),
+      name: profiler.target().name(),
+      id: profiler.target().id()
+    }))));
+    const controller = Timeline.TimelinePanel.instance()._controller;
+    controller.traceEventsCollected([{
```

Comment:
why do we do this?

Reply:
There are two problems with builtin support:
- it is not ready for dead in the middle of profiling targets,
- it has some web related assumption, e.g. expects main target with tracing controller.

By providing events manually we can solve both problems.